codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -509,10 +509,20 @@ class JsonParserGenerator(
                         "Builder" to symbolProvider.symbolForBuilder(shape),
                     )
                     deserializeStructInner(shape.members())
-                    if (hasFallibleBuilder(shape, symbolProvider)) {
-                        rust("Ok(Some(builder.build()?))")
+
+                    // TODO the need for this check should be eliminated or the server code should be updated.
+                    if (codegenContext.target == CodegenTarget.SERVER) {
+                        if (returnSymbolToParse.isUnconstrained) {
+                            rust("Ok(Some(builder))")
+                        } else {
+                            rust("Ok(Some(builder.build()))")
+                        }
                     } else {
-                        rust("Ok(Some(builder.build()))")
+                        if (hasFallibleBuilder(shape, symbolProvider)) {
```

Comment:
I think this can be if `(hasFaillibleBuilder || !returnSymbolToParse.isUnconstrained)`?

Alternatively, can we push `returnSymbolToParse` as an argument to `hasFallibleBuilder`? That might help us make sure we caught all of these locations.